### PR TITLE
Adds feature branch development guidelines

### LINKED ISSUE #117: [PROPOSAL] Feature branch development
### What are you proposing?
This is based on the [original proposal in OpenSearch Dashboards.](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2887) We use a hybrid of the feature branch development model and the Continuous integration development model [1]. We introduce feature branch development with the following safeguards:

1. [High frequency integration](https://martinfowler.com/articles/branching-patterns.html#High-frequencyIntegration): We raise PR's to main not once the feature is complete, but whenever we reach the closest integration point during development. e.g. once the plugin is bootstrapped, or a core feature is complete. we also frequently rebase changes from main back into the feature branch
a. This makes the integration frequency much higher allowing us to catch integration issues much quicker.
b. This still lets collaboratively develop on a big feature that is not ready for main
2. Treat feature branch PR's the same as PR's to main. i.e. CI is run on all PR's, no incomplete work should be merged, Tests are necessary for all PR's etc.
a. This maintains the code quality going into each feature making the integration to main PR's much easier and quicker
b. More visibility during development since it gives reviewers the necessary time to review each PR.
3. Feature specific labels: This helps identify feature related issues and PR's.

[1]Ref: https://martinfowler.com/articles/branching-patterns.html

### What users have asked for this feature?
Many big feature projects for OpenSearch Dashboards have asked to use the feature branch model to collaborate on the main repo and asked for guidance on the correct practices to follow there. e.g. Alerting anywhere, Point in time, Multiple data sources etc. The observability team too has to recently used this model but has run into some of the issues outlined below when using it.

### What problems are you trying to solve?
The feature branch model today is not well defined which allows for different style of this model being used to build features that cause issues down the road like
    1. A massive PR to main that is very hard to accurately review
    2. Code quality suffers because the intermediate PR's aren't reviewed carefully since many of them are work in progress, missteps are ignored as followup tasks that arent tracked.
    3. Breaking changes arent identified until the last minute due to newer changes in main

### What is the developer experience going to be?
The initial attempt will be slightly slower since the initial bootstrapping effort for a feature may take longer to meet the bar for a valid feature branch PR, but subsequent changes should be quicker since we are sure that the changes introduced so far will not break the experience and that we dont have to work around them. It also makes asynchronous collaboration easier.

#### Are there any security considerations? 
No

#### Are there any breaking changes to the API
no

### What is the user experience going to be?
Not user facing

#### Are there breaking changes to the User Experience?
No

### Why should it be built? Any reason not to?
It can slow down the pace of feature development since it has additional restrictions and does not allow for work in progress code. However this tradeoff is not as significant since the developer is always free to make such changes in their local fork and these limitations only apply to the feature branches in the original repo.

### What will it take to execute?
Codifying this proposal in the repo and using it as guidance for other project repo's to follow 

### Any remaining open questions?
No

### PR DESCRIPTION
**Pull request by ashwin-pc**
Signed-off-by: Ashwin Pc <ashwinpc@amazon.com>

### Description
Add feature branch development guidelines to the repo

### Issues Resolved
#117 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CONTRIBUTING.md
Comment (dblock) [inline CONTRIBUTING.md:5]: TOC doesn't match, update.

File: CONTRIBUTING.md
Comment (dblock) [inline CONTRIBUTING.md:138]: Capitalize Like Other Titles to be consistent.

File: CONTRIBUTING.md
Comment (dblock) [inline CONTRIBUTING.md:150]: This is good. 

Run a spell checker, e.g. `corrsponding` is mispelled.
Make sure all sentences have periods at the end, like the first bullet and 2.1.

Consider shortening the text and maybe getting rid of bullet points. What are the most important points? I think the fact that feature branches need to match the process on main, but then you make a lot of assumptions about that process itself. We're not that prescriptive or we don't need to repeat all the things we do on main, saying "Feature branch processes, including but not limited to working through pull requests, doing code reviews and ensuring that CI/CD passes, should be the same as on `main`."

**Review by peternied — COMMENTED** [review]
Love that you are documenting this process for how/when to make feature branches.  Could you add some maintainers from Dashboards to take a look and provide feedback too?

FYI @opensearch-project/opensearch-dashboards-core

File: CONTRIBUTING.md
Comment (joshuarrrr) [inline CONTRIBUTING.md:142]:
```suggestion
1. High frequency integration: PR's to `main` are raised not once the feature is complete, but whenever we reach the closest integration point during development. For example, once the plugin is bootstrapped, or a core feature is complete. We also frequently rebase changes from main back into the feature branch.
```

File: CONTRIBUTING.md
Comment (joshuarrrr) [inline CONTRIBUTING.md:143]:
```suggestion
    1. This makes the integration frequency much higher and allows us to catch integration issues sooner.
```

File: CONTRIBUTING.md
Comment (joshuarrrr) [inline CONTRIBUTING.md:144]: nit - end all list items with a period for consistency

```suggestion
    2. This still lets us collaboratively develop on a big feature that is not yet ready for `main`.
```

File: CONTRIBUTING.md
Comment (joshuarrrr) [inline CONTRIBUTING.md:140]: Do we also want to mention that feature branches should have a meta issue associated with them? And would an example of a feature branch with all these features (meta issue, branch, label) be useful to link to?

File: CONTRIBUTING.md
Comment (joshuarrrr) [inline CONTRIBUTING.md:145]:
```suggestion
2. Treat feature branch PR's the same as PR's to `main`. CI should run on all PR's, no incomplete work should be merged, tests are necessary, etc.
```

What about CHANGELOG enforcement? It seems sensible to make an exception for that (only editing the CHANGELOG as part of the PRs to `main`.

**Review by joshuarrrr — COMMENTED** [review]
A couple of questions, but otherwise just some proofreading suggestions. I also agree with @dblock that we should streamline where possible to avoid being overly prescriptive, although I appreciate all of that detail and clarity at the repo level.

File: CONTRIBUTING.md
Comment (mingshl) [inline CONTRIBUTING.md:140]: Would you please point the direction(might be links) to guide the contributors towards finding what is the best way to ask the maintainers to create the feature branch? Should they create an issue and request maintainers? Please provide more details.

**Comment by peternied** [comment]
@ashwin-pc Once the outstanding feedback has been addressed I will be happy to approval this change.

File: CONTRIBUTING.md
Comment (ashwin-pc) [reply CONTRIBUTING.md:5]: This change actually fixes the TOC for the Feature requests heading. The previous link does not actually work. It was my markdown linter that actually caught this error

File: CONTRIBUTING.md
Comment (ashwin-pc) [reply CONTRIBUTING.md:140]: Yeah i like the idea of a meta issue but do we have a good example yet to link here? Also since the feature branches will be deprecated once the feature is complete, the links will be outdated coon too.

File: CONTRIBUTING.md
Comment (ashwin-pc) [reply CONTRIBUTING.md:150]: I am hesitant on being less prescriptive here since that has often led to more confusion than less. Especially as the number of repositories that we have grows, being more opinionated is quite useful since it removes ambiguity. That being said, I do call out that these are not rules, but rather guidelines, so repo owners can make judgement calls as they see fit.

File: CONTRIBUTING.md
Comment (ashwin-pc) [reply CONTRIBUTING.md:140]: Good callout, i will add a sentence for contributors looking to introduce a new feature

File: CONTRIBUTING.md
Comment (ashwin-pc) [reply CONTRIBUTING.md]: Given the open discussion around changelog and back-porting thats happening right now, i'd keep that out of this until we have clarity there

**Comment by ashwin-pc** [comment]
@peternied @dblock Ive addressed the comments here. Let me know if there are any other concerns you have about the PR.

File: CONTRIBUTING.md
Comment (peternied) [reply CONTRIBUTING.md:5]: Looked at @ashwin-pc's forks seem to work as expected. [[link](https://github.com/ashwin-pc/opensearch.github/blob/doc/feature-branching/CONTRIBUTING.md#feature-requests--proposals)]

File: CONTRIBUTING.md
Comment (peternied) [reply CONTRIBUTING.md:150]: This section looks fine to me, thanks for cleaning up the structure.

**Review by peternied — APPROVED** [review]
Looks good to me @ashwin-pc.

Can we get another review?

**Review by dblock — APPROVED** [review]

**Pull request merged**
